### PR TITLE
180 bug insulate entry classes within alignedtextgrid instances from the global class

### PR DESCRIPTION
**Pull request by JoFrhwld**
This PR will clone entry classes upon creation of an AlignedTextGrid, thus insulating any changes made to those entry classes from the global environment, or other AlignedTextGrid objects.

### CODE REVIEW

**Comment by JoFrhwld** [comment]
Still needs tests. @chrisbrickhouse I think if this is merged and published, it should resolve your problem with fave-syllabify: https://github.com/Forced-Alignment-and-Vowel-Extraction/fave-syllabify/issues/5

**Review by chrisbrickhouse — CHANGES_REQUESTED** [review]
Typo and iteration comments are blockers, but in general lgtm

File: src/aligned_textgrid/aligned_textgrid.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 from aligned_textgrid.sequences.tiers import SequenceTier, TierGroup
 from aligned_textgrid.points.tiers import SequencePointTier, PointsGroup
 from aligned_textgrid.mixins.within import WithinMixins
-from aligned_textgrid.custom_classes import custom_classes
+from aligned_textgrid.custom_classes import custom_classes, clone_class, get_class_hierarhcy
```
[the remark this answers is not in the document]
Comment (JoFrhwld) [reply src/aligned_textgrid/aligned_textgrid.py]: fixed

File: tests/test_aligned_textgrid.py [hunk 2]
```diff
@@ -69,20 +69,22 @@ def test_read_partial(self):
...
-            )      
-        assert len(atg1) == 4
-        assert [len(tg) == 1 for tg in atg1]
+    # def test_read_partial2(self):
+    #     atg1 = AlignedTextGrid(
+    #         textgrid_path="tests/test_data/KY25A_1.TextGrid", 
+    #         entry_classes=[Phone]
+    #         )      
+    #     assert len(atg1) == 4
+    #     assert [len(tg) == 1 for tg in atg1]
```
[the remark this answers is not in the document]
Comment (JoFrhwld) [reply tests/test_aligned_textgrid.py:78]: Good point, I'll remove. The weirdness of the test is reading in a textgrid with an entry class that's not top of the hierarchy. This winds up throwing errors when cloning and `get_class_hierarchy()` is run.

File: tests/test_aligned_textgrid.py [hunk 3]
```diff
@@ -123,7 +125,8 @@ def test_equiv(self):
         for g1, g2 in zip(self.atg1.tier_groups, self.atg2.tier_groups):
             assert g1.tier_names == g2.tier_names
         
-        assert self.atg1.entry_classes == self.atg2.entry_classes
+        # Intentionally broken
```
[the remark this answers is not in the document]
Comment (JoFrhwld) [reply tests/test_aligned_textgrid.py]: Done

File: src/aligned_textgrid/custom_classes.py [hunk 4]
```diff
@@ -151,4 +149,90 @@ def _bottom_constructor(self):
...
+    if issubclass(entry_class, SequencePoint):
+        cloned = type(
+                entry_class.__name__, 
+                (entry_class, ), 
+                dict(entry_class.__dict__)
+        )
+    
+    return cloned
+
+def get_class_hierarhcy(
```
[the remark this answers is not in the document]
Comment (JoFrhwld) [reply src/aligned_textgrid/custom_classes.py]: fixed

File: src/aligned_textgrid/aligned_textgrid.py [hunk 5]
```diff
@@ -175,8 +176,56 @@ def _extend_classes(
...
+            ) -> list[SequenceInterval|SequencePoint]|list[list[SequenceInterval|SequencePoint]]:
+        
+        flat_classes = entry_classes
+        if type(entry_classes[0]) is list:
+            flat_classes = [c for tg in entry_classes for c in tg]
+        
+        unique_classes = list(set(flat_classes))
+
+        points = [c for c in unique_classes if issubclass(c, SequencePoint)]
+        intervals = [c for c in unique_classes if issubclass(c, SequenceInterval)]
```
[the remark this answers is not in the document]
Comment (JoFrhwld) [reply src/aligned_textgrid/aligned_textgrid.py]: Ah, I guess I forgot that `if`s in list comprehensions are evaluated in sequence, and not as if it were `if a & b`

**Pull request merged**